File: src/content/3.7/code/scala/snippet17.scala
implicit val streamFunctor = new Functor[Stream] {
  def fmap[A, B](f: A => B)(fa: Stream[A]): Stream[B] = fa match {
    case Stream(a, as) =>
      Stream(() => f(a()), () => fmap(f)(as()))
  }
}

File: src/content/1.10/code/ocaml/snippet17.ml
module Reader_Functor (T : sig
  type e
end) : Functor with type 'a t = (T.e, 'a) reader = struct
  type 'a t = (T.e, 'a) reader

  let fmap f (Reader g) = Reader (fun e -> f (g e))
end
